File: app/src/main/java/com/blundell/tut/TimeStampFormatter.java
package com.blundell.tut;

import java.util.Date;
import java.util.concurrent.TimeUnit;

import org.joda.time.DateTime;
import org.joda.time.Days;
import org.joda.time.Hours;
import org.joda.time.Minutes;
import org.joda.time.Months;
import org.joda.time.Weeks;
import org.joda.time.Years;

class TimeStampFormatter {

    /**
     * For use with java.util.Date
     */
    public String format(Date timestamp) {
        long millisFromNow = getMillisFromNow(timestamp);

        long minutesFromNow = TimeUnit.MILLISECONDS.toMinutes(millisFromNow);
        if (minutesFromNow < 1) {
            return "just now";
        }
        long hoursFromNow = TimeUnit.MILLISECONDS.toHours(millisFromNow);
        if (hoursFromNow < 1) {
            return formatMinutes(minutesFromNow);
        }
        long daysFromNow = TimeUnit.MILLISECONDS.toDays(millisFromNow);
        if (daysFromNow < 1) {
            return formatHours(hoursFromNow);
        }
        long weeksFromNow = TimeUnit.MILLISECONDS.toDays(millisFromNow) / 7;
        if (weeksFromNow < 1) {
            return formatDays(daysFromNow);
        }
        long monthsFromNow = TimeUnit.MILLISECONDS.toDays(millisFromNow) / 30;
        if (monthsFromNow < 1) {
            return formatWeeks(weeksFromNow);
        }
        long yearsFromNow = TimeUnit.MILLISECONDS.toDays(millisFromNow) / 365;
        if (yearsFromNow < 1) {
            return formatMonths(monthsFromNow);
        }
        return formatYears(yearsFromNow);
    }

    /**
     * For use with org.joda.DateTime
     */
    public String format(DateTime commentedAt) {
        DateTime now = DateTime.now();
        Minutes minutesBetween = Minutes.minutesBetween(commentedAt, now);
        if (minutesBetween.isLessThan(Minutes.ONE)) {
            return "just now";
        }
        Hours hoursBetween = Hours.hoursBetween(commentedAt, now);
        if (hoursBetween.isLessThan(Hours.ONE)) {
            return formatMinutes(minutesBetween.getMinutes());
        }
        Days daysBetween = Days.daysBetween(commentedAt, now);
        if (daysBetween.isLessThan(Days.ONE)) {
            return formatHours(hoursBetween.getHours());
        }
        Weeks weeksBetween = Weeks.weeksBetween(commentedAt, now);
        if (weeksBetween.isLessThan(Weeks.ONE)) {
            return formatDays(daysBetween.getDays());
        }
        Months monthsBetween = Months.monthsBetween(commentedAt, now);
        if (monthsBetween.isLessThan(Months.ONE)) {
            return formatWeeks(weeksBetween.getWeeks());
        }
        Years yearsBetween = Years.yearsBetween(commentedAt, now);
        if (yearsBetween.isLessThan(Years.ONE)) {
            return formatMonths(monthsBetween.getMonths());
        }
        return formatYears(yearsBetween.getYears());
    }

    private long getMillisFromNow(Date commentedAt) {
        long commentedAtMillis = commentedAt.getTime();
        long nowMillis = System.currentTimeMillis();
        return nowMillis - commentedAtMillis;
    }

    private String formatMinutes(long minutes) {
        return format(minutes, " minute ago", " minutes ago");
    }

    private String formatHours(long hours) {
        return format(hours, " hour ago", " hours ago");
    }

    private String formatDays(long days) {
        return format(days, " day ago", " days ago");
    }

    private String formatWeeks(long weeks) {
        return format(weeks, " week ago", " weeks ago");
    }

    private String formatMonths(long months) {
        return format(months, " month ago", " months ago");
    }

    private String formatYears(long years) {
        return format(years, " year ago", " years ago");
    }

    private String format(long hand, String singular, String plural) {
        if (hand == 1) {
            return hand + singular;
        } else {
            return hand + plural;
        }
    }

}

File: app/src/main/java/com/blundell/tut/TimeStampFormatter.kt
package com.blundell.tut

import java.util.Date
import java.util.concurrent.TimeUnit

import org.joda.time.DateTime
import org.joda.time.Days
import org.joda.time.Hours
import org.joda.time.Minutes
import org.joda.time.Months
import org.joda.time.Weeks
import org.joda.time.Years

class TimeStampFormatter {

    /**
     * For use with java.util.Date
     */
    fun format(timestamp: Date?): String {
        val millisFromNow = getMillisFromNow(timestamp)

        val minutesFromNow = TimeUnit.MILLISECONDS.toMinutes(millisFromNow)
        if (minutesFromNow < 1) {
            return "just now"
        }
        val hoursFromNow = TimeUnit.MILLISECONDS.toHours(millisFromNow)
        if (hoursFromNow < 1) {
            return formatMinutes(minutesFromNow)
        }
        val daysFromNow = TimeUnit.MILLISECONDS.toDays(millisFromNow)
        if (daysFromNow < 1) {
            return formatHours(hoursFromNow)
        }
        val weeksFromNow = TimeUnit.MILLISECONDS.toDays(millisFromNow) / 7
        if (weeksFromNow < 1) {
            return formatDays(daysFromNow)
        }
        val monthsFromNow = TimeUnit.MILLISECONDS.toDays(millisFromNow) / 30
        if (monthsFromNow < 1) {
            return formatWeeks(weeksFromNow)
        }
        val yearsFromNow = TimeUnit.MILLISECONDS.toDays(millisFromNow) / 365
        return if (yearsFromNow < 1) {
            formatMonths(monthsFromNow)
        } else formatYears(yearsFromNow)
    }

    private fun getMillisFromNow(commentedAt: Date?): Long {
        val commentedAtMillis = commentedAt?.time
        val nowMillis = System.currentTimeMillis()
        return nowMillis - commentedAtMillis!!
    }

    /**
     * For use with org.joda.DateTime
     */
    fun format(commentedAt: DateTime): String {
        val now = DateTime.now()
        val minutesBetween = Minutes.minutesBetween(commentedAt, now)
        if (minutesBetween.isLessThan(Minutes.ONE)) {
            return "just now"
        }
        val hoursBetween = Hours.hoursBetween(commentedAt, now)
        if (hoursBetween.isLessThan(Hours.ONE)) {
            return formatMinutes(minutesBetween.minutes.toLong())
        }
        val daysBetween = Days.daysBetween(commentedAt, now)
        if (daysBetween.isLessThan(Days.ONE)) {
            return formatHours(hoursBetween.hours.toLong())
        }
        val weeksBetween = Weeks.weeksBetween(commentedAt, now)
        if (weeksBetween.isLessThan(Weeks.ONE)) {
            return formatDays(daysBetween.days.toLong())
        }
        val monthsBetween = Months.monthsBetween(commentedAt, now)
        if (monthsBetween.isLessThan(Months.ONE)) {
            return formatWeeks(weeksBetween.weeks.toLong())
        }
        val yearsBetween = Years.yearsBetween(commentedAt, now)
        return if (yearsBetween.isLessThan(Years.ONE)) {
            formatMonths(monthsBetween.months.toLong())
        } else formatYears(yearsBetween.years.toLong())
    }

    private fun formatMinutes(minutes: Long): String {
        return format(minutes, " minute ago", " minutes ago")
    }

    private fun formatHours(hours: Long): String {
        return format(hours, " hour ago", " hours ago")
    }

    private fun formatDays(days: Long): String {
        return format(days, " day ago", " days ago")
    }

    private fun formatWeeks(weeks: Long): String {
        return format(weeks, " week ago", " weeks ago")
    }

    private fun formatMonths(months: Long): String {
        return format(months, " month ago", " months ago")
    }

    private fun formatYears(years: Long): String {
        return format(years, " year ago", " years ago")
    }

    private fun format(hand: Long, singular: String, plural: String): String {
        return if (hand == 1L) {
            hand.toString() + singular
        } else {
            hand.toString() + plural
        }
    }

}
